### PR TITLE
reduce memory usage around ngt implementation & refactor agent/lb & auto-generate unit test 

### PR DESCRIPTION
**Pull request by kpango**
### Description:
This pr includes 9 features.

1. Eliminated the use of ebuf for shared memory, and implemented error buffer pooling for proprietary memory using sync.Pool package .
internal/core/algorithm/ngt/ngt.go
internal/core/algorithm/ngt/option.go

2. Changed to implement more strict error handling in Search and SearchByID.
pkg/agent/core/ngt/handler/grpc/handler.go

3. Users can now know more about the search vector of why a search failed during a SearchByID Operation.
4. Insert on CreateIndex now retries only once if the error is not ErrIncompatibleDimensionSize.
5. Automatic adjustment of the pool size has been added. CreateIndex now allocates a more appropriate size of memory.
pkg/agent/core/ngt/service/ngt.go

6. The vqueue's reference operation used to carry Global Lock, so it has been changed to Read Lock.
pkg/agent/core/ngt/service/vqueue/queue.go

7. Float is now cached and reused.
pkg/gateway/lb/handler/grpc/handler.go

8. re-autogenerate test code (please ignore them)
9. replace go.uber.org/goleak to internal/goleak  (please ignore them)

### How Has This Been Tested?:
I have tested it on GithubActions E2E testing & tested in k3d on my local linux environment.

### Environment:

- Go Version: 1.17.1
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.12.3

### Types of changes:

- [x] Bug fix [type/bug]
- [x] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [x] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/core/algorithm/ngt/ngt.go [hunk 1]
```diff
@@ -210,12 +210,22 @@ func gen(isLoad bool, opts ...Option) (NGT, error) {
...
+		},
+	}
+
+	for i := 0; i < 20; i++ {
+		n.PutErrorBuffer(C.ngt_create_error_object())
+	}
 
-	n.prop = C.ngt_create_property(n.ebuf)
+	ebuf := n.GetErrorBuffer()
+	n.prop = C.ngt_create_property(ebuf)
```
Comment (kevindiu) [inline internal/core/algorithm/ngt/ngt.go:224]: can we also cache this? 🤔

Comment (kpango) [reply internal/core/algorithm/ngt/ngt.go:224]: I already cached this buffer please read GetErrorBuffer and PutErrorBuffer codes.

File: internal/core/algorithm/ngt/ngt.go [hunk 2]
```diff
@@ -310,11 +328,12 @@ func (n *ngt) Search(vec []float32, size int, epsilon, radius float32) (result [
 		return nil, errors.ErrIncompatibleDimensionSize(len(vec), int(n.dimension))
 	}
 
-	results := C.ngt_create_empty_results(n.ebuf)
+	ebuf := n.GetErrorBuffer()
+	results := C.ngt_create_empty_results(ebuf)
```
Comment (kevindiu) [inline internal/core/algorithm/ngt/ngt.go:332]: do you think we can also cache this empty result object?

File: pkg/agent/core/ngt/handler/grpc/handler.go [hunk 3]
```diff
@@ -189,7 +189,8 @@ func (s *server) Search(ctx context.Context, req *payload.Search_Request) (res *
 				})
 			log.Debug(err)
 			stat = trace.StatusCodeAborted(err.Error())
-		case errors.Is(err, errors.ErrEmptySearchResult):
+		case errors.Is(err, errors.ErrEmptySearchResult),
+			err == nil && res == nil:
```
Comment (kevindiu) [inline pkg/agent/core/ngt/handler/grpc/handler.go:193]: why do we add this checking?

File: pkg/agent/core/ngt/handler/grpc/handler.go [hunk 4]
```diff
@@ -1117,18 +1185,20 @@ func (s *server) Upsert(ctx context.Context, req *payload.Upsert_Request) (loc *
 		loc, err = s.Update(ctx, &payload.Update_Request{
 			Vector: req.GetVector(),
 			Config: &payload.Update_Config{
-				Timestamp: req.GetConfig().GetTimestamp(),
+				Timestamp:            req.GetConfig().GetTimestamp(),
```
Comment (kevindiu) [inline pkg/agent/core/ngt/handler/grpc/handler.go:1188]: is there any chance that the req or req.GetConfig() is nil?

File: pkg/agent/core/ngt/handler/grpc/handler.go (hunk 3 above)
Comment (kpango) [reply pkg/agent/core/ngt/handler/grpc/handler.go:193]: some time calling CGO (internal/core/algo/ngt) returns no error and empty result, so I added and modified it to return NotFound status.
Maybe it's caused by creating indeices.

File: pkg/agent/core/ngt/handler/grpc/handler.go (hunk 4 above)
Comment (kpango) [reply pkg/agent/core/ngt/handler/grpc/handler.go:1188]: It's protocol buffer auto generated code, req.GetConfig() never panics if req is nil so req.GetConfig().GetTimestamp() never panics if req is nil or req.GetConfig()  is nil, please read code.

File: internal/core/algorithm/ngt/ngt.go (hunk 2 above)
Comment (kpango) [reply internal/core/algorithm/ngt/ngt.go:332]: Hmmm, that's an interesting idea, but I think it's better to do it some other time as we have to check many things like search accuracy, RaceCondition, etc.
For ebufs, there is a clear way to destroy, but for result containers, there is no such I/F, so we have to consider how to destroy used buffers.

**Review by kevindiu — APPROVED** [review]

**Pull request merged**